### DISCUSSION TITLE
Best way to use ServiceClient with Depency Injection in ASP.Net Core?

### DISCUSSION THREAD

**Discussion by burkenyo** [microsoft/PowerPlatform-DataverseServiceClient#129] [Q&A]
I would like to be able to inject a `ServiceClient` instance into my services used in an ASP.Net Core app. I have been able to do so using the `ActivatorUtilities` class thus far:

```
    // inside my public void ConfigureServices(IServiceCollection services) method
    services.AddScoped<ServiceClient>(services =>
    {
        var baseUri = _config["Dataverse:BaseUri"];
        var clientId = _config["Dataverse:ClientId"];
        var clientSecret = _config["Dataverse:ClientSecret"];

        var connectionString = $@"
            AuthType=ClientSecret;
            Url={baseUri};
            ClientId={clientId};
            ClientSecret={clientSecret}
        ";

        return ActivatorUtilities.CreateInstance<ServiceClient>(services, connectionString);
    });
```

My biggest question is which service lifetime I should use. Currently I am using Scoped. Should I use singleton?

**Comment by BetimBeja** [comment]
Please read Matt's answer here:
https://github.com/microsoft/PowerPlatform-DataverseServiceClient/issues/78#issuecomment-745414318

It should be the same for the asp.net core as for the azure functions.

**Comment by heroes3lover** [comment]
Hello @burkenyo,
How is going with your choice? currently I got same question as well.
I have registered the ServiceClient as scoped, but I notice occasionally Web App returns 500 error with "DataverseConnectionException: Failed to connect to Dataverse" meanwhile some other ServiceClient still succeed with the operations.
I suspect the authentication failed during new ServiceClient initialized when too many requests come in. 
I got maybe 30K requests in one hour.
